### ISSUE TITLE
Does not build under PHP8 RC 1

### ISSUE THREAD

**Issue by SoftCreatR** [swoole/swoole-src#3717]
I am currently working on a Dockerfile, that uses `php:8.0.0beta4-cli-alpine` and Swoole (from master) builds just fine.

However, I tried to switch to `php:8.0.0rc1-cli-alpine`, but building Swoole fails:

```
creating libtool
appending configuration tag "CXX" to libtool
checking for ld used by g++... /usr/x86_64-alpine-linux-musl/bin/ld
checking if the linker (/usr/x86_64-alpine-linux-musl/bin/ld) is GNU ld... yes
checking whether the g++ linker (/usr/x86_64-alpine-linux-musl/bin/ld) supports shared libraries... yes
checking for g++ option to produce PIC... -fPIC
checking if g++ PIC flag -fPIC works... yes
checking if g++ static flag -static works... yes
checking if g++ supports -c -o file.o... yes
checking whether the g++ linker (/usr/x86_64-alpine-linux-musl/bin/ld) supports shared libraries... yes
checking dynamic linker characteristics... GNU/Linux ld.so
(cached) (cached) checking how to hardcode library paths into programs... immediate
configure: patching config.h.in
configure: creating ./config.status
config.status: creating config.h
/bin/sh /swoole-src/libtool --mode=compile g++ -I. -I/swoole-src -I/swoole-src/include -I/swoole-src/main -I/swoole-src -I/usr/local/include/php -I/usr/local/include/php/main -I/usr/local/include/php/TSRM -I/usr/local/include/php/Zend -I/usr/local/include/php/ext -I/usr/local/include/php/ext/date/lib -I/swoole-src -I/swoole-src/include -I/swoole-src/thirdparty/hiredis  -DHAVE_CONFIG_H  -g -O2 -Wall -Wno-unused-function -Wno-deprecated -Wno-deprecated-declarations -std=c++11    -c /swoole-src/php_swoole.cc -o php_swoole.lo
mkdir .libs
 g++ -I. -I/swoole-src -I/swoole-src/include -I/swoole-src/main -I/swoole-src -I/usr/local/include/php -I/usr/local/include/php/main -I/usr/local/include/php/TSRM -I/usr/local/include/php/Zend -I/usr/local/include/php/ext -I/usr/local/include/php/ext/date/lib -I/swoole-src -I/swoole-src/include -I/swoole-src/thirdparty/hiredis -DHAVE_CONFIG_H -g -O2 -Wall -Wno-unused-function -Wno-deprecated -Wno-deprecated-declarations -std=c++11 -c /swoole-src/php_swoole.cc  -fPIC -DPIC -o .libs/php_swoole.o
In file included from /swoole-src/php_swoole_cxx.h:19,
                 from /swoole-src/php_swoole.cc:16:
/swoole-src/php_swoole.h: In function 'char* php_swoole_http_build_query(zval*, size_t*, smart_str*)':
/swoole-src/php_swoole.h:1132:121: error: no match for 'operator==' (operand types are 'void' and 'ZEND_RESULT_CODE')
 1132 |     if (php_url_encode_hash_ex(HASH_OF(zdata), formstr, NULL, 0, NULL, 0, NULL, 0, NULL, NULL, (int) PHP_QUERY_RFC1738) == FAILURE)
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^~ ~~~~~~~
      |                               |                                                                                            |
      |                               void                                                                                         ZEND_RESULT_CODE
make: *** [Makefile:209: php_swoole.lo] Error 1
```

Well, I also tried to build from the `alpine` branch, but the result is worse:

```
creating libtool
appending configuration tag "CXX" to libtool
checking for ld used by g++... /usr/x86_64-alpine-linux-musl/bin/ld
checking if the linker (/usr/x86_64-alpine-linux-musl/bin/ld) is GNU ld... yes
checking whether the g++ linker (/usr/x86_64-alpine-linux-musl/bin/ld) supports shared libraries... yes
checking for g++ option to produce PIC... -fPIC
checking if g++ PIC flag -fPIC works... yes
checking if g++ static flag -static works... yes
checking if g++ supports -c -o file.o... yes
checking whether the g++ linker (/usr/x86_64-alpine-linux-musl/bin/ld) supports shared libraries... yes
checking dynamic linker characteristics... GNU/Linux ld.so
(cached) (cached) checking how to hardcode library paths into programs... immediate
configure: patching config.h.in
configure: creating ./config.status
config.status: creating config.h
/bin/sh /swoole-src/libtool --mode=compile g++ -I. -I/swoole-src -I/swoole-src/include -I/swoole-src/main -I/swoole-src -I/usr/local/include/php -I/usr/local/include/php/main -I/usr/local/include/php/TSRM -I/usr/local/include/php/Zend -I/usr/local/include/php/ext -I/usr/local/include/php/ext/date/lib -I/swoole-src -I/swoole-src/include -I/swoole-src/thirdparty/hiredis  -DHAVE_CONFIG_H  -g -O2 -Wall -Wno-unused-function -Wno-deprecated -Wno-deprecated-declarations -std=c++11    -c /swoole-src/php_swoole_cxx.cc -o php_swoole_cxx.lo
mkdir .libs
 g++ -I. -I/swoole-src -I/swoole-src/include -I/swoole-src/main -I/swoole-src -I/usr/local/include/php -I/usr/local/include/php/main -I/usr/local/include/php/TSRM -I/usr/local/include/php/Zend -I/usr/local/include/php/ext -I/usr/local/include/php/ext/date/lib -I/swoole-src -I/swoole-src/include -I/swoole-src/thirdparty/hiredis -DHAVE_CONFIG_H -g -O2 -Wall -Wno-unused-function -Wno-deprecated -Wno-deprecated-declarations -std=c++11 -c /swoole-src/php_swoole_cxx.cc  -fPIC -DPIC -o .libs/php_swoole_cxx.o
In file included from /swoole-src/php_swoole_cxx.h:19,
                 from /swoole-src/php_swoole_cxx.cc:1:
/swoole-src/php_swoole.h: In function 'int add_assoc_ulong_safe_ex(zval*, const char*, size_t, zend_ulong)':
/swoole-src/php_swoole.h:712:33: error: void value not ignored as it ought to be
  712 |         return add_assoc_long_ex(arg, key, key_len, value);
      |                ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
/swoole-src/php_swoole.h:718:36: error: void value not ignored as it ought to be
  718 |         return add_assoc_stringl_ex(arg, key, key_len, buf, len);
      |                ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/swoole-src/php_swoole.h: In function 'zval* sw_zend_read_property(zend_class_entry*, zval*, const char*, int, int)':
/swoole-src/php_swoole.h:923:49: error: cannot convert 'zval*' {aka '_zval_struct*'} to 'zend_object*' {aka '_zend_object*'}
  923 |     zval rv, *property = zend_read_property(ce, obj, s, len, silent, &rv);
      |                                                 ^~~
      |                                                 |
      |                                                 zval* {aka _zval_struct*}
In file included from /usr/local/include/php/main/php.h:35,
                 from /swoole-src/php_swoole.h:25,
                 from /swoole-src/php_swoole_cxx.h:19,
                 from /swoole-src/php_swoole_cxx.cc:1:
/usr/local/include/php/Zend/zend_API.h:404:73: note:   initializing argument 2 of 'zval* zend_read_property(zend_class_entry*, zend_object*, const char*, size_t, zend_bool, zval*)'
  404 | ZEND_API zval *zend_read_property(zend_class_entry *scope, zend_object *object, const char *name, size_t name_length, zend_bool silent, zval *rv);
      |                                                            ~~~~~~~~~~~~~^~~~~~
In file included from /swoole-src/php_swoole_cxx.h:19,
                 from /swoole-src/php_swoole_cxx.cc:1:
/swoole-src/php_swoole.h:926:39: error: cannot convert 'zval*' {aka '_zval_struct*'} to 'zend_object*' {aka '_zend_object*'}
  926 |         zend_update_property_null(ce, obj, s, len);
      |                                       ^~~
      |                                       |
      |                                       zval* {aka _zval_struct*}
In file included from /usr/local/include/php/main/php.h:35,
                 from /swoole-src/php_swoole.h:25,
                 from /swoole-src/php_swoole_cxx.h:19,
                 from /swoole-src/php_swoole_cxx.cc:1:
/usr/local/include/php/Zend/zend_API.h:385:79: note:   initializing argument 2 of 'void zend_update_property_null(zend_class_entry*, zend_object*, const char*, size_t)'
  385 | ZEND_API void zend_update_property_null(zend_class_entry *scope, zend_object *object, const char *name, size_t name_length);
      |                                                                  ~~~~~~~~~~~~~^~~~~~
In file included from /swoole-src/php_swoole_cxx.h:19,
                 from /swoole-src/php_swoole_cxx.cc:1:
/swoole-src/php_swoole.h:927:39: error: cannot convert 'zval*' {aka '_zval_struct*'} to 'zend_object*' {aka '_zend_object*'}
  927 |         return zend_read_property(ce, obj, s, len, silent, &rv);
      |                                       ^~~
      |                                       |
      |                                       zval* {aka _zval_struct*}
In file included from /usr/local/include/php/main/php.h:35,
                 from /swoole-src/php_swoole.h:25,
                 from /swoole-src/php_swoole_cxx.h:19,
                 from /swoole-src/php_swoole_cxx.cc:1:
/usr/local/include/php/Zend/zend_API.h:404:73: note:   initializing argument 2 of 'zval* zend_read_property(zend_class_entry*, zend_object*, const char*, size_t, zend_bool, zval*)'
  404 | ZEND_API zval *zend_read_property(zend_class_entry *scope, zend_object *object, const char *name, size_t name_length, zend_bool silent, zval *rv);
      |                                                            ~~~~~~~~~~~~~^~~~~~
In file included from /swoole-src/php_swoole_cxx.h:19,
                 from /swoole-src/php_swoole_cxx.cc:1:
/swoole-src/php_swoole.h: In function 'zval* sw_zend_read_property_not_null(zend_class_entry*, zval*, const char*, int, int)':
/swoole-src/php_swoole.h:934:49: error: cannot convert 'zval*' {aka '_zval_struct*'} to 'zend_object*' {aka '_zend_object*'}
  934 |     zval rv, *property = zend_read_property(ce, obj, s, len, silent, &rv);
      |                                                 ^~~
      |                                                 |
      |                                                 zval* {aka _zval_struct*}
In file included from /usr/local/include/php/main/php.h:35,
                 from /swoole-src/php_swoole.h:25,
                 from /swoole-src/php_swoole_cxx.h:19,
                 from /swoole-src/php_swoole_cxx.cc:1:
/usr/local/include/php/Zend/zend_API.h:404:73: note:   initializing argument 2 of 'zval* zend_read_property(zend_class_entry*, zend_object*, const char*, size_t, zend_bool, zval*)'
  404 | ZEND_API zval *zend_read_property(zend_class_entry *scope, zend_object *object, const char *name, size_t name_length, zend_bool silent, zval *rv);
      |                                                            ~~~~~~~~~~~~~^~~~~~
In file included from /swoole-src/php_swoole_cxx.h:19,
                 from /swoole-src/php_swoole_cxx.cc:1:
/swoole-src/php_swoole.h: In function 'zval* sw_zend_update_and_read_property_array(zend_class_entry*, zval*, const char*, int)':
/swoole-src/php_swoole.h:943:30: error: cannot convert 'zval*' {aka '_zval_struct*'} to 'zend_object*' {aka '_zend_object*'}
  943 |     zend_update_property(ce, obj, s, len, &ztmp);
      |                              ^~~
      |                              |
      |                              zval* {aka _zval_struct*}
In file included from /usr/local/include/php/main/php.h:35,
                 from /swoole-src/php_swoole.h:25,
                 from /swoole-src/php_swoole_cxx.h:19,
                 from /swoole-src/php_swoole_cxx.cc:1:
/usr/local/include/php/Zend/zend_API.h:384:74: note:   initializing argument 2 of 'void zend_update_property(zend_class_entry*, zend_object*, const char*, size_t, zval*)'
  384 | ZEND_API void zend_update_property(zend_class_entry *scope, zend_object *object, const char *name, size_t name_length, zval *value);
      |                                                             ~~~~~~~~~~~~~^~~~~~
In file included from /swoole-src/php_swoole_cxx.h:19,
                 from /swoole-src/php_swoole_cxx.cc:1:
/swoole-src/php_swoole.h:945:35: error: cannot convert 'zval*' {aka '_zval_struct*'} to 'zend_object*' {aka '_zend_object*'}
  945 |     return zend_read_property(ce, obj, s, len, 1, &ztmp);
      |                                   ^~~
      |                                   |
      |                                   zval* {aka _zval_struct*}
In file included from /usr/local/include/php/main/php.h:35,
                 from /swoole-src/php_swoole.h:25,
                 from /swoole-src/php_swoole_cxx.h:19,
                 from /swoole-src/php_swoole_cxx.cc:1:
/usr/local/include/php/Zend/zend_API.h:404:73: note:   initializing argument 2 of 'zval* zend_read_property(zend_class_entry*, zend_object*, const char*, size_t, zend_bool, zval*)'
  404 | ZEND_API zval *zend_read_property(zend_class_entry *scope, zend_object *object, const char *name, size_t name_length, zend_bool silent, zval *rv);
      |                                                            ~~~~~~~~~~~~~^~~~~~
In file included from /swoole-src/php_swoole_cxx.h:19,
                 from /swoole-src/php_swoole_cxx.cc:1:
/swoole-src/php_swoole.h: In function 'zval* sw_zend_read_and_convert_property_array(zend_class_entry*, zval*, const char*, int, int)':
/swoole-src/php_swoole.h:950:49: error: cannot convert 'zval*' {aka '_zval_struct*'} to 'zend_object*' {aka '_zend_object*'}
  950 |     zval rv, *property = zend_read_property(ce, obj, s, len, silent, &rv);
      |                                                 ^~~
      |                                                 |
      |                                                 zval* {aka _zval_struct*}
In file included from /usr/local/include/php/main/php.h:35,
                 from /swoole-src/php_swoole.h:25,
                 from /swoole-src/php_swoole_cxx.h:19,
                 from /swoole-src/php_swoole_cxx.cc:1:
/usr/local/include/php/Zend/zend_API.h:404:73: note:   initializing argument 2 of 'zval* zend_read_property(zend_class_entry*, zend_object*, const char*, size_t, zend_bool, zval*)'
  404 | ZEND_API zval *zend_read_property(zend_class_entry *scope, zend_object *object, const char *name, size_t name_length, zend_bool silent, zval *rv);
      |                                                            ~~~~~~~~~~~~~^~~~~~
In file included from /swoole-src/php_swoole_cxx.h:19,
                 from /swoole-src/php_swoole_cxx.cc:1:
/swoole-src/php_swoole.h: In function 'int sw_zend_call_function_ex(zval*, zend_fcall_info_cache*, uint32_t, zval*, zval*)':
/swoole-src/php_swoole.h:1054:9: error: 'zend_fcall_info' {aka 'struct _zend_fcall_info'} has no member named 'no_separation'
 1054 |     fci.no_separation = 0;
      |         ^~~~~~~~~~~~~
/swoole-src/php_swoole.h: In function 'char* php_swoole_http_build_query(zval*, size_t*, smart_str*)':
/swoole-src/php_swoole.h:1195:121: error: no match for 'operator==' (operand types are 'void' and 'ZEND_RESULT_CODE')
 1195 |     if (php_url_encode_hash_ex(HASH_OF(zdata), formstr, NULL, 0, NULL, 0, NULL, 0, NULL, NULL, (int) PHP_QUERY_RFC1738) == FAILURE)
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^~ ~~~~~~~
      |                               |                                                                                            |
      |                               void                                                                                         ZEND_RESULT_CODE
/swoole-src/php_swoole_cxx.cc: In function 'bool zend::eval(std::string, std::string)':
/swoole-src/php_swoole_cxx.cc:61:30: error: invalid conversion from 'zend_op_array* (*)(zend_string*, const char*)' {aka '_zend_op_array* (*)(_zend_string*, const char*)'} to 'zend_op_array* (*)(zval*, const char*)' {aka '_zend_op_array* (*)(_zval_struct*, const char*)'} [-fpermissive]
   61 |         old_compile_string = zend_compile_string;
      |                              ^~~~~~~~~~~~~~~~~~~
      |                              |
      |                              zend_op_array* (*)(zend_string*, const char*) {aka _zend_op_array* (*)(_zend_string*, const char*)}
/swoole-src/php_swoole_cxx.cc:64:27: error: invalid conversion from 'zend_op_array* (*)(zval*, const char*)' {aka '_zend_op_array* (*)(_zval_struct*, const char*)'} to 'zend_op_array* (*)(zend_string*, const char*)' {aka '_zend_op_array* (*)(_zend_string*, const char*)'} [-fpermissive]
   64 |     zend_compile_string = swoole_compile_string;
      |                           ^~~~~~~~~~~~~~~~~~~~~
      |                           |
      |                           zend_op_array* (*)(zval*, const char*) {aka _zend_op_array* (*)(_zval_struct*, const char*)}
/swoole-src/php_swoole_cxx.cc:67:27: error: invalid conversion from 'zend_op_array* (*)(zval*, const char*)' {aka '_zend_op_array* (*)(_zval_struct*, const char*)'} to 'zend_op_array* (*)(zend_string*, const char*)' {aka '_zend_op_array* (*)(_zend_string*, const char*)'} [-fpermissive]
   67 |     zend_compile_string = old_compile_string;
      |                           ^~~~~~~~~~~~~~~~~~
      |                           |
      |                           zend_op_array* (*)(zval*, const char*) {aka _zend_op_array* (*)(_zval_struct*, const char*)}
make: *** [Makefile:209: php_swoole_cxx.lo] Error 1
```

**Comment by SoftCreatR** [comment]
Heads up: Seems to be fixed via #3713

**Comment by doubaokun** [comment]
@SoftCreatR merged to the master branch.
